sympy/diffgeom/diffgeom.py

```diff
@@ -1084,6 +1086,374 @@ def __call__(self, field):
+        except KeyError:
+            try:
+                systems = list(T.atoms(CoordSystem))
+                if len(systems) > 1:
+                    raise NotImplemented
+                self.CoordSystem = systems.pop()
+            except AttributeError: #Scalar case (probably)
+                raise NotImplementedError("Specify coordinate system with coordinate_system=...")
+        self.coords = self.CoordSystem.coord_functions()
+        self.base_oneforms = self.CoordSystem.base_oneforms()
+        self.base_vectors = self.CoordSystem.base_vectors()
+        self.n = len(self.coords)
+        self.covariant_order=covariant_order(T)
+        self.contravariant_order=contravariant_order(T)
+        self.order = self.covariant_order+self.contravariant_order
+        self.shape = [self.n for _ in range(self.order)]
+        self.indices = list(itertools.product(range(self.n),repeat=self.order))
+        self.contravariant_slots=self._contravariant_slots(T)
+        self.covariant_slots=list(range(self.order))
+        [self.covariant_slots.remove(i) for i in self.contravariant_slots]
+        def basishelper(p,k):
+            if self.contravariant_slots.count(k) == 0:
+                return self.base_vectors[p[k]]
+            else:
+                return self.coords[p[k]]
+        Tbases = [
+            (p,[basishelper(list(p),k) for k in range(self.order)]) for p in self.indices
+        ]
+        if self.order==0: #scalar case
+            self.tensor = MutableSparseNDimArray(T)
+        else:
+            self.tensor = MutableSparseNDimArray.zeros(*self.shape)
+            for (p,exI) in Tbases:
+                Tval = T.rcall(*exI)
+                if Tval != 0:
+                    self.tensor[p]=Tval
+                else:
+                    self.indices.remove(p)
+    def _init_from_array(self,component_array,variance,coordinate_system):
+        self.CoordSystem = coordinate_system
+        self.coords = self.CoordSystem.coord_functions()
+        self.base_oneforms = self.CoordSystem.base_oneforms()
+        self.base_vectors = self.CoordSystem.base_vectors()
+        self.n = len(self.coords)
+        self.order = len(variance)
+        self.shape = [self.n for _ in range(self.order)]
+        self.covariant_slots=[k for k in range(self.order) if variance[k]==1]
+        self.covariant_order = len(self.covariant_slots)
+        self.contravariant_slots=[k for k in range(self.order) if variance[k]==-1]
+        self.contravariant_order = len(self.contravariant_slots)
+        self.indices = list(itertools.product(range(self.n),repeat=self.order))
+        self.tensor = MutableSparseNDimArray.zeros(*self.shape)
+        component_array = MutableSparseNDimArray(component_array)
+        for index in self.indices:
+            if component_array[index]!=0:
+                self.tensor[index] = component_array[index]
+            else:
+                self.indices.remove(index)
+    def __init__(self, *args, **kwargs):
```

Comment:
This signature looks too general to be really helpful. I would prefer something like the signature of `_init_from_array` with named parameters. For initialization from tensors, there could be a public function that would extract the parameters from a tensor and return the TensorArray object created with those parameters. That could be, for example, `tensorarray(T)`.

Reply:
If I understand your comment, I think this is already implemented in the TensorArray class.  TensorArray(T) takes a tensor (in the sense of the diffgeom module), and converts it to a multidimensional array internally (see Example 1).  But with named parameters, we can convert a NDimArray to a TensorArray (see Example 2).

Reply:
I think that the current signature is too general. I would split it into two parts (code for which already exists). One part, `_init_from_array`, would become `__init__`, and `_init_from_tensor` would be made an independent function that would create the TensorArray object via its new `__init__`. Then the docstring for TensorArray would document the parameters (i.e. array, variance and coordsys). Most of the code is already there, It only should be reorganized.

Reply:
Ok, I've added a longer docstring.  The primary invocation of the class is by way of _init_from_tensor, not _init_from_array, so I do not agree with this proposed reorganization.  Both initializations already are independent.